### PR TITLE
Use specific assertions

### PR DESCRIPTION
**Pull request by CYBAI**
Similar to #19865 
r? jdm 

Note: Should I squash all the commits into one commit?

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because it should not break anything

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19868)

### CODE REVIEW

File: components/compositing/compositor.rs [hunk 1]
```diff
@@ -495,6 +495,8 @@ impl<Window: WindowMethods> IOCompositor<Window> {
             }
 
             (Msg::IsReadyToSaveImageReply(is_ready), ShutdownState::NotShuttingDown) => {
+                // TODO: We cannot change this to `assert_eq` because `ReadyState` doesn't
+                //       derive `Debug`
                 assert!(self.ready_to_save_state == ReadyState::WaitingForConstellationReply);
```
Comment (emilio) [inline components/compositing/compositor.rs]: I think it should derive `Debug`, mind changing it or filing an issue for that?

**Review by emilio — APPROVED** [review]
Looks good! If you feel like it, `cd components; rg 'assert!' | rg '!=' ` shows a couple of other assertions that can be changed to `assert_ne!` instead.

File: components/compositing/compositor.rs (hunk 1 above)
Comment (CYBAI) [reply components/compositing/compositor.rs]: Let me derive it in this PR! Thanks!

**Comment by CYBAI** [comment]
@emilio Great! Let me update for `assert_ne!` as well! Thanks for the great catch!

**Comment by CYBAI** [comment]
There're still some `assert!(a == b)` and one `assert!(a != b)` in the codebase with above patches.

They're all related to `#[dom_struct]`.
If I tried to change `assert!` to `assert_eq!`; then compiler will complain that I should derive `Debug` for them.
However, if I tried to derive `Debug` for the struct, compiler will complain that I also need to derive `Debug` for those types used in the attribute of the struct.

Take the last negative `assert!` as an example,

https://github.com/servo/servo/blob/47b3693d7564b32813746c1dbb6ef29dbb08ffaf/components/script/dom/attr.rs#L231

When I change the line to

```rust
assert_ne!(old.get_attribute(&ns, &self.identifier.local_name).r(), Some(self))
```

Then, the compile result will be 

```rust
error[E0277]: the trait bound `dom::attr::Attr: std::fmt::Debug` is not satisfied
   --> components/script/dom/attr.rs:231:17
    |
231 |                 assert_ne!(old.get_attribute(&ns, &self.identifier.local_name).r(), Some(self))
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `dom::attr::Attr` cannot be formatted using `:?`; if it is defined in your crate, add `#[derive(Debug)]` or manually implement it
    |
    = help: the trait `std::fmt::Debug` is not implemented for `dom::attr::Attr`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `&dom::attr::Attr`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `std::option::Option<&dom::attr::Attr>`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `&std::option::Option<&dom::attr::Attr>`
    = note: required by `std::fmt::Debug::fmt`
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0277]: the trait bound `dom::attr::Attr: std::fmt::Debug` is not satisfied
   --> components/script/dom/attr.rs:231:17
    |
231 |                 assert_ne!(old.get_attribute(&ns, &self.identifier.local_name).r(), Some(self))
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `dom::attr::Attr` cannot be formatted using `:?`; if it is defined in your crate, add `#[derive(Debug)]` or manually implement it
    |
    = help: the trait `std::fmt::Debug` is not implemented for `dom::attr::Attr`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `&dom::attr::Attr`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `std::option::Option<&dom::attr::Attr>`
    = note: required because of the requirements on the impl of `std::fmt::Debug` for `&std::option::Option<&dom::attr::Attr>`
    = note: required by `std::fmt::Debug::fmt`
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 2 previous errors

error: Could not compile `script`.
```

Then, I just tried to derive `Debug` for `Attr` struct,

```diff
// https://dom.spec.whatwg.org/#interface-attr
#[dom_struct]
+#[derive(Debug)]
pub struct Attr {
    reflector_: Reflector,
    identifier: AttrIdentifier,
    value: DomRefCell<AttrValue>,

    /// the element that owns this attribute.
    owner: MutNullableDom<Element>,
}
```

the compile result will be

```rust
error[E0277]: the trait bound `dom::bindings::reflector::Reflector: std::fmt::Debug` is not satisfied
  --> components/script/dom/attr.rs:31:5
   |
31 |     reflector_: Reflector,
   |     ^^^^^^^^^^^^^^^^^^^^^ `dom::bindings::reflector::Reflector` cannot be formatted using `:?`; if it is defined in your crate, add `#[derive(Debug)]` or manually implement it
   |
   = help: the trait `std::fmt::Debug` is not implemented for `dom::bindings::reflector::Reflector`
   = note: required because of the requirements on the impl of `std::fmt::Debug` for `&dom::bindings::reflector::Reflector`
   = note: required for the cast to the object type `std::fmt::Debug`

error[E0277]: the trait bound `dom::bindings::root::MutNullableDom<dom::element::Element>: std::fmt::Debug` is not satisfied
  --> components/script/dom/attr.rs:36:5
   |
36 |     owner: MutNullableDom<Element>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `dom::bindings::root::MutNullableDom<dom::element::Element>` cannot be formatted using `:?`; if it is defined in your crate, add `#[derive(Debug)]` or manually implement it
   |
   = help: the trait `std::fmt::Debug` is not implemented for `dom::bindings::root::MutNullableDom<dom::element::Element>`
   = note: required because of the requirements on the impl of `std::fmt::Debug` for `&dom::bindings::root::MutNullableDom<dom::element::Element>`
   = note: required for the cast to the object type `std::fmt::Debug`

error: aborting due to 2 previous errors

error: Could not compile `script`.
```

Should I derive `Debug` for those related type?
If yes, I'd like to file an issue for that and open another PR to solve it!

**Comment by emilio** [comment]
>  If yes, I'd like to file an issue for that and open another PR to solve it!

Maybe yes, but I wouldn't worry too much about it. This looks good to me, thanks!

@bors-servo r+

**Pull request merged**